File: examples/ts/Scripts/GuideScene.ts
/// <reference path="../fairygui.d.ts" />

import { Callable, Rect2, Vector2 } from "godot";
import { DemoSceneBase } from "./DemoSceneBase";

export default class GuideScene extends DemoSceneBase {
    private _guideLayer: GComponent | null = null;
    private _bagBtn: GuiObject | null = null;

    protected ContinueInit(): void {
        UIPackage.addPackage("res://Resources/UI/Guide");

        this._view = UIPackage.createObject("Guide", "Main");
        this._groot!.addChild(this._view);

        this._guideLayer = UIPackage.createObject("Guide", "GuideLayer");
        if (this._guideLayer != null) {
            this._guideLayer.makeFullScreen();
            this._guideLayer.addRelation(this._groot!, FguiRelationType.RightRight, false);
        }

        this._bagBtn = this._view.getChild("bagBtn");
        if (this._bagBtn != null) {
            this._bagBtn.addClickListener(
                Callable.create(() => {
                    if (this._guideLayer != null && this._guideLayer.getParent() != null) {
                        this._guideLayer.removeFromParent();
                    }
                }),
            );
        }

        const n2 = this._view.getChild("n2");
        if (n2 != null) {
            n2.addClickListener(Callable.create(this._showGuide.bind(this)));
        }
    }

    private _showGuide(): void {
        if (this._guideLayer == null || this._bagBtn == null) {
            return;
        }
        this._groot!.addChild(this._guideLayer);

        const rect = this._bagBtn.transformRect(
            new Rect2(0, 0, this._bagBtn.getWidth(), this._bagBtn.getHeight()),
            this._guideLayer,
        );
        const window = this._guideLayer.getChild("window");
        if (window != null) {
            window.setSize(rect.size.x, rect.size.y);
            GTweenHelper.getInstance()
                .toVec2(window.getPosition(), rect.position, 0.5)
                .setTarget(window, GTweener.TweenPropType.POSITION);
        }
    }
}

File: examples/csharp/Scripts/GuideScene.cs
using Godot;

namespace FairyGUI.Examples;

public partial class GuideScene : DemoSceneBase
{
    private GuiObject _guideLayer;
    private GuiObject _bagBtn;

    protected override void ContinueInit()
    {
        UIPackage.AddPackage("res://Resources/UI/Guide");

        _view = UIPackage.CreateObject("Guide", "Main");
        _groot.AddChild(_view);

        _guideLayer = UIPackage.CreateObject("Guide", "GuideLayer");
        if (_guideLayer != null)
        {
            _guideLayer.MakeFullScreen();
            _guideLayer.AddRelation(_groot, GuiObject.RelationType.RightRight, false);
        }

        _bagBtn = _view.GetChild("bagBtn");
        _bagBtn?.AddClickListener(Callable.From(() =>
        {
            if (_guideLayer != null && _guideLayer.GetParent() != null)
                _guideLayer.RemoveFromParent();
        }));

        var n2 = _view.GetChild("n2");
        n2?.AddClickListener(new Callable(this, MethodName.ShowGuide));
    }

    private void ShowGuide()
    {
        if (_guideLayer == null || _bagBtn == null)
            return;
        _groot.AddChild(_guideLayer);

        var rect = _bagBtn.TransformRect(
            new Rect2(Vector2.Zero, new Vector2(_bagBtn.GetWidth(), _bagBtn.GetHeight())),
            _guideLayer);
        var window = _guideLayer.GetChild("window");
        if (window != null)
        {
            window.SetSize(rect.Size.X, rect.Size.Y);
            GTweenHelper.GetInstance()
                .ToVec2(window.GetPosition(), rect.Position, 0.5f)
                .SetTarget(window, (int)GTweener.TweenPropType.Position);
        }
    }
}
